example/grammar/magnet.cpp

```diff
@@ -0,0 +1,496 @@
+};
+
+struct magnet_link_rule_t
+{
+    using value_type = magnet_link_view;
+
+    // Specs:
+    // https://www.bittorrent.org/beps/bep_0005.html
+    // https://www.bittorrent.org/beps/bep_0009.html
+    // https://www.bittorrent.org/beps/bep_0053.html
+    // https://github.com/webtorrent/magnet-uri
+    // https://en.wikipedia.org/wiki/Magnet_URI_scheme
+
+    result< value_type >
+    parse( char const*& it, char const* end ) const noexcept
+    {
+        // 1) Check if string starts with url
+        // (absolute_uri_rule_t should return url_view)
+        char const* it0 = it;
+        result<absolute_uri_rule_t::value_type> rrv =
+            grammar::parse(it, end, absolute_uri_rule);
+        if(!rrv)
+        {
+            it = it0;
+            return error::invalid;
+        }
+
+        // 2) Parse this url
+        magnet_link_view m;
+        m.u_ = *parse_absolute_uri({it0, it});
+
+        // 3) Check if exact topics are valid urls
+        // and that we have at least one
+        auto ps = m.u_.encoded_params();
+        auto pit = ps.begin();
+        auto pend = ps.end();
+        auto f = magnet_link_view::is_exact_topic{};
+        pit = std::find_if(pit, pend, f);
+        if (pit == pend)
+        {
+            // no topic in the magnet link
+            return error::invalid;
+        }
+
+        // all topics parse as valid urls
+        if (!std::all_of(pit, pend, [](
+            query_param_encoded_view p)
+        {
+            if (!magnet_link_view::is_exact_topic{}(p))
+                return true;
+            result<url_view> u = parse_uri(p.value);
+            return u.has_value();
+        }))
+            return error::invalid;
+
+        // all other fields are optional
+        // magnet link is OK
+        return m;
+    }
+};
+
+constexpr magnet_link_rule_t magnet_link_rule{};
+
+// user-facing function
+/** Return a parsed magnet link from a string, or error.
+*/
+result< magnet_link_view >
+parse_magnet_link( string_view s ) noexcept
+{
+    return grammar::parse(s, magnet_link_rule);
+}
+
+
+int main(int argc, char** argv)
+{
+    // This example shows how to use custom parsing
+    // to process alternate URI schemes, in this
+    // case "magnet"
+    if (argc != 2) {
+        std::cout << argv[0] << "\n";
+        std::cout << "magnet <link>\n"
+                     "example: magnet magnet:?xt=urn:btih:d2474e86c95b19b8bcfdb92bc12c9d44667cfa36"
+                                            "&dn=Leaves+of+Grass+by+Walt+Whitman.epub"
+                                            "&tr=udp%3A%2F%2Ftracker.example4.com%3A80"
+                                            "&tr=udp%3A%2F%2Ftracker.example5.com%3A80"
+                                            "&tr=udp%3A%2F%2Ftracker.example3.com%3A6969"
+                                            "&tr=udp%3A%2F%2Ftracker.example2.com%3A80"
+                                            "&tr=udp%3A%2F%2Ftracker.example1.com%3A1337\n";
```

Comment:
Are these links to copyrighted material?

Reply:
Nop...

display name: Leaves of Grass by Walt Whitman.epub

Leaves of grass by Whitman, Walt, 1819-1892

https://archive.org/search.php?query=creator%3A%22Whitman%2C+Walt%2C+1819-1892%22